modules/flist/flist.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"crypto/md5"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zosv2/modules"
+)
+
+const (
+	defaultStorage = "zdb://hub.grid.tf:9900"
+	defaultRoot    = "/var/modules/flist"
+)
+
+type flistModule struct {
+	// root directory where all
+	// the working file of the module will be located
+	root string
+
+	// underneath are the path for each
+	// sub folder used by the flist module
+	flist      string
+	cache      string
+	mountpoint string
+	pid        string
+	log        string
+}
+
+// New creates a new flistModule
+func New(root string) modules.Flister {
+	if root == "" {
+		root = defaultRoot
+	}
+
+	// prepare directory layout for the module
+	for _, path := range []string{"flist", "cache", "mountpoint", "pid", "log"} {
+		if err := os.MkdirAll(filepath.Join(root, path), 0770); err != nil {
+			panic(err)
+		}
+	}
+
+	return &flistModule{
+		root:       root,
+		flist:      filepath.Join(root, "flist"),
+		cache:      filepath.Join(root, "cache"),
+		mountpoint: filepath.Join(root, "mountpoint"),
+		pid:        filepath.Join(root, "pid"),
+		log:        filepath.Join(root, "log"),
+	}
+}
+
+// Mount implements the Flister.Mount interface
+func (f *flistModule) Mount(url, storage string) (string, error) {
+	sublog := log.With().Str("url", url).Str("storage", storage).Logger()
+	sublog.Info().Msg("request to mount flist")
+
+	if storage == "" {
+		storage = defaultStorage
+	}
+
+	rc, err := downloadFlist(url)
+	if err != nil {
+		sublog.Err(err).Msg("fail to download flist")
+		return "", err
+	}
+	defer rc.Close()
+
+	flistPath, err := f.saveFlist(rc)
+	if err != nil {
+		sublog.Err(err).Msg("fail to save flist on disk")
+		return "", err
+	}
+
+	rnd, err := random()
+	if err != nil {
+		sublog.Error().Err(err).Msg("fail to generate random id for the mount")
+		return "", err
+	}
+	mountpoint := filepath.Join(f.mountpoint, rnd)
+	if err := os.MkdirAll(mountpoint, 0770); err != nil {
+		return "", err
+	}
+	pidPath := filepath.Join(f.pid, rnd) + ".pid"
+	logPath := filepath.Join(f.log, rnd) + ".log"
+
+	args := []string{
+		"-backend", filepath.Join(f.root, "backend", rnd),
+		"-cache", f.cache,
+		"-meta", flistPath,
+		"-storage-url", storage,
+		"-daemon",
```

Comment:
`daemon` Is that a new flag? I can't find it in g8ufs implementation. Also do we really need to demonize g8ufs ? We can always just spawn the fs, then wait on both the pid file and that the mountpoint is ready.

In current 0-core the way we do it is that we wait for the "mount ready" on fuse stdout, then we know for sure that the mount is ready. I think this approach is better than waiting for 2 seconds

Reply:
> daemon Is that a new flag? I can't find it in g8ufs implementation. Also do we really need to demonize g8ufs ? We can always just spawn the fs, then wait on both the pid file and that the mountpoint is ready.

I believe the have to demonize cause otherwise the g8ufs processes are stopped if the flist module is restarted.

> In current 0-core the way we do it is that we wait for the "mount ready" on fuse stdout, then we know for sure that the mount is ready. I think this approach is better than waiting for 2 seconds

Yep this is something I need to improve, there is a todo in the code for it.